### PR TITLE
SOC on SD Card

### PR DESCRIPTION
**Pull request by DJ90864**
### Changelist 
Moved to SD to store SOC values
CRC peripheral enabled for SOC data integrity

### Testing Done
SD card operations are tested with different SOC values on Quadruna
Custom Reset needs to be tested since there was an issue with the CAN Rx signal

### CODE REVIEW

File: firmware/quadruna/BMS/src/app/app_soc.c [hunk 1]
```diff
@@ -180,3 +202,66 @@ void app_soc_resetSocCustomValue(float soc_percent)
...
+    }
+}
+
+void app_soc_writeSocToSd(float soc)
+{
+    uint8_t sd_write_data[HW_DEVICE_SECTOR_SIZE];
+    convert_float_to_bytes(sd_write_data, soc);
+    uint32_t soc_value = ARRAY_TO_UINT32(sd_write_data);
+
+#ifdef TARGET_EMBEDDED
```
Comment (will-chaba) [inline firmware/quadruna/BMS/src/app/app_soc.c:259]: Why not wrap the entire function? Is there a testable benefit to leaving the above 3 lines in?

**Review by Jiajun-Huang — CHANGES_REQUESTED** [review]

File: firmware/shared/src/hw/hw_sd.c [hunk 2]
```diff
@@ -3,21 +3,38 @@
...
+bool hw_sd_checkSdPresent(void)
+{
+    return !hw_gpio_readPin(&sd->sd_present);
+}
 
-    sd = sd_in;
+void hw_sd_markInitComplete(void)
+{
+    sd->sd_init_complete = true;
+}
```
[the remark this answers is not in the document]
Comment (DJ90864) [reply firmware/shared/src/hw/hw_sd.c:33]: done!

File: firmware/quadruna/BMS/src/cubemx/Src/main.c [hunk 3]
```diff
@@ -451,6 +491,40 @@ static void MX_FDCAN1_Init(void)
...
+    hsd1.Init.BusWide             = SDMMC_BUS_WIDE_4B;
+    hsd1.Init.HardwareFlowControl = SDMMC_HARDWARE_FLOW_CONTROL_ENABLE;
+    hsd1.Init.ClockDiv            = 9;
+    if (HAL_SD_Init(&hsd1) != HAL_OK)
+    {
+        Error_Handler();
+    }
+    /* USER CODE BEGIN SDMMC1_Init 2 */
+    sd_inited = true;
+    HAL_SD_ConfigWideBusOperation(&hsd1, SDMMC_BUS_WIDE_4B);
```
[the remark this answers is not in the document]
Comment (DJ90864) [reply firmware/quadruna/BMS/src/cubemx/Src/main.c:524]: done!

File: firmware/shared/src/hw/hw_sd.c
[the remark this answers is not in the document]
Comment (DJ90864) [reply firmware/shared/src/hw/hw_sd.c:1]: makes sense!
I moved it from hw_sd to app_soc instead.

File: firmware/shared/src/hw/hw_sd.h [hunk 4]
```diff
@@ -1,6 +1,7 @@
 #pragma once
 
 #include "hw_hal.h"
+#include "hw_gpio.h"
```
Comment (Jiajun-Huang) [inline firmware/shared/src/hw/hw_sd.h:4]: Can you remove the gpio stuff from hw_sd?

File: firmware/shared/src/hw/hw_sd.h [hunk 5]
```diff
@@ -23,9 +26,23 @@ typedef enum
...
+ * @param sd_config SD Card Configuration
+ */
+void hw_sd_init(SdCard *sd_config);
 
-void hw_sd_init(const SdCard *sd);
+/**
+ * @brief Check for presence of SD card
+ * @return true if SD card present
+ */
+bool hw_sd_checkSdPresent(void);
```
Comment (Jiajun-Huang) [inline firmware/shared/src/hw/hw_sd.h:39]: remove here

File: firmware/shared/src/hw/hw_sd.h [hunk 6]
```diff
@@ -23,9 +26,23 @@ typedef enum
...
+
+/**
+ * @brief Check if SD is ready for operation
+ * @return true if initialization complete and SD present
+ */
+bool hw_sd_sdReady(void);
```
Comment (Jiajun-Huang) [inline firmware/shared/src/hw/hw_sd.h:45]: also here

File: firmware/shared/src/hw/hw_sd.c [hunk 7]
```diff
@@ -3,21 +3,28 @@
...
-        return;
+        sd->sd_init_complete = true;
     }
+}
 
-    sd = sd_in;
+bool hw_sd_sdReady(void)
+{
+    return sd->sd_init_complete && !hw_gpio_readPin(&sd_present);
 }
```
Comment (Jiajun-Huang) [inline firmware/shared/src/hw/hw_sd.c:28]: I guess here is also not needed

**Review by Jiajun-Huang — CHANGES_REQUESTED** [review]
Just rm the GPIO  stuff from hw_sd. Others are good!

**Review by Jiajun-Huang — APPROVED** [review]

File: firmware/quadruna/BMS/src/app/app_soc.c [hunk 8]
```diff
@@ -79,48 +119,35 @@ float app_soc_getOcvFromSoc(float soc_percent)
...
+    {
+        if (IS_IN_RANGE(0.0f, SERIES_ELEMENT_FULL_CHARGE_C * 1.25f, saved_soc_c))
+        {
+            stats.charge_c   = (double)saved_soc_c;
+            stats.is_corrupt = false;
+        }
+    }
+    else
+    {
+        // anything to do if fail to read it corrrectly from the SD?
```
Comment (will-chaba) [inline firmware/quadruna/BMS/src/app/app_soc.c:142]: Everything that needs to be done is done already, set SOC corrupt and set SOC to -1 to signify it needs to be reset based on cell voltage

**Review by will-chaba — APPROVED** [review]
Looks good to me! Tested functional on BMS May 5, 2024

**Pull request merged**